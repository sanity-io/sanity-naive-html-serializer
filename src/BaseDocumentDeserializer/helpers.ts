import {htmlToBlocks} from '@sanity/block-tools'
import {Schema} from '@sanity/schema'
import {ObjectField, PortableTextSpan, PortableTextTextBlock} from 'sanity'

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [{type: 'block'}],
        },
      ],
    },
  ],
})

export const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

export const noSchemaWarning = (obj: Element): string =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

//helper to handle messy input -- take advantage
//of blockTools' sanitizing behavior for single strings
export const preprocess = (html: string): string => {
  const intermediateBlocks = htmlToBlocks(
    `<p>${html}</p>`,
    blockContentType
  ) as PortableTextTextBlock<PortableTextSpan>[]
  if (!intermediateBlocks.length) {
    throw new Error(`Error parsing string '${html}'`)
  }
  return intermediateBlocks[0].children[0].text
}
