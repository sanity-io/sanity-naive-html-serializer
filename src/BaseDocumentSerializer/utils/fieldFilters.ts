import {ObjectField, TypedObject} from 'sanity'

const META_FIELDS = ['_key', '_type', '_id']

/*
 * Helper. If field-level translation pattern used, only sends over
 * content from the base language. Works recursively, so if users
 * use this pattern several layers deep, base language fields will still be found.
 */
export const languageObjectFieldFilter = (
  obj: Record<string, any>,
  baseLang: string
): Record<string, any> => {
  const filterToLangField = (childObj: Record<string, any>) => {
    const filteredObj: Record<string, any> = {}
    filteredObj[baseLang] = childObj[baseLang]
    META_FIELDS.forEach((field) => {
      if (childObj[field]) {
        filteredObj[field] = childObj[field]
      }
    })
    return filteredObj
  }

  const findBaseLang = (childObj: Record<string, any>): Record<string, any> => {
    const filteredObj: Record<string, any> = {}
    META_FIELDS.forEach((field) => {
      if (childObj[field]) {
        filteredObj[field] = childObj[field]
      }
    })

    for (const key in childObj) {
      if (childObj.hasOwnProperty(key)) {
        const value: any = childObj[key]
        //we've reached a base language field, add it to
        //what we want to send to translation
        if (value.hasOwnProperty(baseLang)) {
          filteredObj[key] = filterToLangField(value)
        }
        //we have an array that may have language fields in its objects
        else if (Array.isArray(value) && value.length && typeof value[0] === 'object') {
          //recursively find and filter for any objects that have the base language
          const validLangObjects = value.reduce((validArr, objInArray) => {
            if (objInArray._type === 'block') {
              validArr.push(objInArray)
            } else if (objInArray.hasOwnProperty(baseLang)) {
              validArr.push(filterToLangField(objInArray))
            } else {
              const filtered = findBaseLang(objInArray)
              const nonMetaFields = Object.keys(filtered).filter(
                (objInArrayKey) => META_FIELDS.indexOf(objInArrayKey) === -1
              )
              if (nonMetaFields.length) {
                validArr.push(filtered)
              }
            }
            return validArr
          }, [])
          if (validLangObjects.length) {
            filteredObj[key] = validLangObjects
          }
        }
        //we have an object nested in an object
        //recurse down the tree
        else if (typeof value === 'object') {
          const nestedLangObj = findBaseLang(value)
          const nonMetaFields = Object.keys(nestedLangObj).filter(
            (nestedObjKey) => META_FIELDS.indexOf(nestedObjKey) === -1
          )
          if (nonMetaFields.length) {
            filteredObj[key] = nestedLangObj
          }
        }
      }
    }
    return filteredObj
  }

  //send top level object into recursive function
  return findBaseLang(obj)
}

/*
 * Eliminates stop-types and non-localizable fields
 * for document-level translation.
 */
export const fieldFilter = (
  obj: Record<string, any>,
  objFields: ObjectField[],
  stopTypes: string[]
): TypedObject => {
  const filteredObj: TypedObject = {_type: obj._type}

  const fieldFilterFunc = (field: Record<string, any>) => {
    if (field.localize === false) {
      return false
    } else if (field.type === 'string' || field.type === 'text') {
      return true
    } else if (Array.isArray(obj[field.name])) {
      return true
    } else if (!stopTypes.includes(field.type)) {
      return true
    }
    return false
  }

  const validFields = [
    ...META_FIELDS,
    ...objFields?.filter(fieldFilterFunc)?.map((field) => field.name),
  ]
  validFields.forEach((field) => {
    if (obj[field]) {
      filteredObj[field] = obj[field]
    }
  })
  return filteredObj
}
