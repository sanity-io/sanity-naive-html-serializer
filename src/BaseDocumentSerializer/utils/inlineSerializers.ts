import {PortableTextComponents} from '@portabletext/to-html'
import {ArbitraryTypedObject} from '@portabletext/types'
import {PortableTextSpan, PortableTextTextBlock, Schema} from 'sanity'
import {BaseDocumentSerializer} from '..'

type ObjectSerializerMethod = ({value}: {value: ArbitraryTypedObject}) => string
type MarkSerializerMethod = ({
  value,
  markType,
  children,
}: {
  value: ArbitraryTypedObject
  markType: string
  children: any[]
}) => string

const validateInlineContent = (
  obj: ArbitraryTypedObject | PortableTextSpan,
  serializers: PortableTextComponents,
  stopTypes: string[]
) =>
  //don't serialize blocks or spans
  obj._type !== 'block' &&
  obj._type !== 'span' &&
  //don't serialize any types that have been explicitly marked as stop types
  !stopTypes.includes(obj._type) &&
  //don't override custom serialization methods from the user
  !serializers.types?.[obj._type]

export const createInlineSerializerTypes = (
  block: PortableTextTextBlock,
  stopTypes: string[],
  serializers: PortableTextComponents,
  schemas: Schema
): PortableTextComponents => {
  const inlineObjects = block.children.filter((child) =>
    validateInlineContent(child, serializers, stopTypes)
  )
  const types: string[] = Array.from(new Set(inlineObjects.map((obj) => obj._type)))
  const serializerFunc: ObjectSerializerMethod = ({value}: {value: ArbitraryTypedObject}) => {
    const wrapperStart = [
      `<span class="default-serialized-object"`,
      `id="${value._key ?? value._id}"`,
      `data-type="${value._type}">`,
    ].join(' ')
    const serializedData = BaseDocumentSerializer(schemas).serializeObject(
      value,
      stopTypes,
      serializers
    )
    const wrapperEnd = `</span>`
    return [wrapperStart, serializedData, wrapperEnd].join('')
  }
  return types.reduce<Record<string, ObjectSerializerMethod>>((acc, type) => {
    acc[type as string] = serializerFunc
    return acc
  }, {})
}

export const createMarkSerializerTypes = (
  block: PortableTextTextBlock,
  stopTypes: string[],
  serializers: PortableTextComponents,
  schemas: Schema
): PortableTextComponents => {
  const marks = block.markDefs?.filter((mark) =>
    validateInlineContent(mark, serializers, stopTypes)
  )
  const markTypes = Array.from(new Set(marks?.map((mark) => mark._type)))
  const serializerFunc = ({
    value,
    markType,
    children,
  }: {
    value: ArbitraryTypedObject
    markType: string
    children: any[]
  }) => {
    const wrapperStart = [
      `<span class="default-serialized-annotation"`,
      `id="${value._key ?? value._id}"`,
      `data-type="${markType}">`,
    ].join(' ')
    const serializedChildren = `<span class="default-serialized-children">${children}</span>`
    const serializedData = BaseDocumentSerializer(schemas).serializeObject(
      value,
      stopTypes,
      serializers
    )
    const wrapperEnd = `</span>`
    return [wrapperStart, serializedChildren, serializedData, wrapperEnd].join('')
  }
  return markTypes.reduce<Record<string, MarkSerializerMethod>>((acc, type) => {
    acc[type as string] = serializerFunc
    return acc
  }, {})
}
