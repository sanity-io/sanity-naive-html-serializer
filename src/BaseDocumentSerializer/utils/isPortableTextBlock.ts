import {ArbitraryTypedObject} from '@portabletext/types'
import {PortableTextTextBlock} from 'sanity'

export const isPortableTextBlock = (el: ArbitraryTypedObject): el is PortableTextTextBlock => {
  return el._type === 'block'
}
