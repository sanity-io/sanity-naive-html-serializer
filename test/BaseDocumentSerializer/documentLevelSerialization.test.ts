import {PortableTextBlock} from 'sanity'
import {getSerialized, getValidFields, toPlainText} from '../helpers'
import {documentLevelArticle, findByClass, getHTMLNode} from './utils'

const serialized = getSerialized(documentLevelArticle, 'document')
const docTree = getHTMLNode(serialized).body.children[0]

test('Global test of working doc-level functionality and snapshot match', () => {
  expect(serialized).toMatchSnapshot()
})
/*
 * Top-level plain text
 */
test('String and text types get serialized correctly at top-level', () => {
  const HTMLString = findByClass(docTree.children, 'title')
  const HTMLText = findByClass(docTree.children, 'snippet')
  expect(HTMLString?.innerHTML).toEqual(documentLevelArticle.title)
  expect(HTMLText?.innerHTML).toEqual(documentLevelArticle.snippet)
})

/*
 * Presence and accuracy of fields
 */
describe('Presence and accuracy of fields in "vanilla" deserialization -- objects', () => {
  //parent node is always div with classname of field with a nested div
  //that has classname of obj type
  const configObj = findByClass(docTree.children, 'config')
  const objectField = configObj!.children[0]

  test('Top-level nested objects contain all serializable fields -- document level', () => {
    const fieldNames = getValidFields(documentLevelArticle.config)
    const foundFieldNames = Array.from(objectField!.children).map((child) => child.className)
    expect(foundFieldNames.sort()).toEqual(fieldNames.sort())
  })

  test('Nested object in object contains all serializable fields -- document level', () => {
    const nestedObject = findByClass(objectField!.children, 'objectAsField')!.children[0]
    const fieldNames = getValidFields(documentLevelArticle.config.objectAsField)
    const foundFieldNames = Array.from(nestedObject!.children).map((child) => child.className)
    expect(foundFieldNames.sort()).toEqual(fieldNames.sort())
  })

  test('Nested object contains accurate values -- document level', () => {
    const title = documentLevelArticle.config.title
    const blockText = toPlainText(documentLevelArticle.config.nestedArrayField)

    expect(objectField?.innerHTML).toContain(title)
    expect(objectField?.innerHTML).toContain(blockText)
  })

  test('Nested object in an object contains accurate values -- document level', () => {
    const nestedObject = findByClass(objectField!.children, 'objectAsField')!.children[0]
    const title = documentLevelArticle.config.objectAsField.title
    const blockText = toPlainText(documentLevelArticle.config.objectAsField.content)

    expect(nestedObject.innerHTML).toContain(title)
    expect(nestedObject.innerHTML).toContain(blockText)
  })
})

describe('Presence and accuracy of fields in vanilla deserialization -- arrays', () => {
  const arrayField = findByClass(docTree.children, 'content')

  test('Array contains all serializable blocks with keys, in order -- document level', () => {
    const origKeys = documentLevelArticle.content.map((block: PortableTextBlock) => block._key)
    const serializedKeys = Array.from(arrayField!.children).map((block) => block.id)
    expect(serializedKeys).toEqual(origKeys)
  })

  test('Array contains top-level block text -- document level', () => {
    const blockText = toPlainText(documentLevelArticle.content).trim()
    const blockStrings = blockText.split('\n\n')
    blockStrings.forEach((substring: string) => expect(arrayField?.innerHTML).toContain(substring))
  })

  test('Object in array contains all serializable fields -- document level', () => {
    const objectInArray = findByClass(arrayField!.children, 'objectField')
    const fieldNames = getValidFields(
      documentLevelArticle.content.find(
        (block: Record<string, any>) => block._type === 'objectField'
      )
    )
    const foundFieldNames = Array.from(objectInArray!.children).map((child) => child.className)
    expect(foundFieldNames.sort()).toEqual(fieldNames.sort())
  })

  test('Object in array contains accurate values in nested object -- document level', () => {
    const objectInArray = findByClass(arrayField!.children, 'objectField')
    const nestedObject = findByClass(objectInArray!.children, 'objectAsField')
    const title = documentLevelArticle.content.find(
      (block: Record<string, any>) => block._type === 'objectField'
    ).objectAsField.title
    const blockText = toPlainText(
      documentLevelArticle.content.find(
        (block: Record<string, any>) => block._type === 'objectField'
      ).objectAsField.content
    ).trim()
    expect(nestedObject?.innerHTML).toContain(title)
    expect(nestedObject?.innerHTML).toContain(blockText)
  })
})
