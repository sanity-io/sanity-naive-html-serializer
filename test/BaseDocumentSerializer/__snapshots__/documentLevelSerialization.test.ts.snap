// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global test of working doc-level functionality and snapshot match 1`] = `
Object {
  "content": "<html><head><meta name=\\"_id\\" content=\\"drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a\\"><meta name=\\"_type\\" content=\\"documentLevelArticle\\"><meta name=\\"_rev\\" content=\\"fxcnqj-uew-gr7-v6r-z965tfm69\\"><meta name=\\"version\\" content=\\"3\\"></head><body><div class=\\"documentLevelArticle\\" id=\\"drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a\\" data-type=\\"object\\"><span class=\\"title\\">My Document-Level Article</span><span class=\\"snippet\\">This is text in my text field.</span><div class=\\"tags\\" data-type=\\"array\\"><span>tag 1</span><span>tag 2</span><span>tag 3</span></div><div class=\\"config\\" data-level=\\"field\\"><div class=\\"objectField\\" data-type=\\"object\\"><span class=\\"title\\">This is an object nested in a document</span><div class=\\"objectAsField\\" data-level=\\"field\\"><div class=\\"childObjectField\\" data-type=\\"object\\"><span class=\\"title\\">This is one level deeper</span><div class=\\"content\\" data-type=\\"array\\"><p id=\\"1aa228ac848b\\">This is a block text 2 levels deep</p></div></div></div><div class=\\"nestedArrayField\\" data-type=\\"array\\"><p id=\\"4a58adc7c507\\">This is block text 1 level deep</p></div></div></div><div class=\\"content\\" data-type=\\"array\\"><p id=\\"a4d177e92666\\">This is block text at the top level.</p><div class=\\"objectField\\" id=\\"c0313627775e\\" data-type=\\"object\\"><span class=\\"title\\">This is an object in top-level block text.</span><div class=\\"objectAsField\\" data-level=\\"field\\"><div class=\\"childObjectField\\" data-type=\\"object\\"><span class=\\"title\\">This is a nested object in an object in top-level block text.</span><div class=\\"content\\" data-type=\\"array\\"><p id=\\"4f24f6fbfae7\\">This is block text in a nested object in an object in top-level block text.</p></div></div></div></div><h1 id=\\"9e2ab13c6d63\\">This is h1 text</h1><h2 id=\\"297142fbaf21\\">This is h2 text</h2><ul id=\\"76e648fc3845\\"><li id=\\"76e648fc3845\\" data-level=\\"1\\">Bullet 1</li></ul><ul id=\\"d090cd8b27d2\\"><li id=\\"d090cd8b27d2\\" data-level=\\"2\\">nested bullet a</li></ul><ul id=\\"1cdffa5d50f5\\"><li id=\\"1cdffa5d50f5\\" data-level=\\"1\\"><h2 id=\\"1cdffa5d50f5\\">Styled bullet 2</h2></li></ul><ol id=\\"da29f5063059\\"><li id=\\"da29f5063059\\" data-level=\\"1\\"><h3 id=\\"da29f5063059\\">Number 1</h3></li></ol></div></div></body></html>",
  "name": "drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a",
}
`;
